### PR TITLE
make git optional, make revision and remote mandatory

### LINKED ISSUE #36: Make Git.remote and Git.revision mandatory fields.
**Is your feature request related to a problem? Please describe.**

Some optional fields should be mandatory and vice versa.

**Describe the solution you'd like**

```typescript
export type CiEnvironment = {
  name: string
  url: string
  buildNumber: string
  git?: Git
}

export type Git = {
  remote: string
  revision: string
  branch?: string
  tag?: string
}
```

If `remote` or `revision` cannot be determined, don't create a `Git` object at all. Those two fields are the bare minimum for useful git information.

### PR DESCRIPTION
**Pull request by davidjgoss**
# Motivation & context

Fixes #36.

## Type of change

- Refactoring/debt (improvement to code design or tooling without changing behaviour)
- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

## Implementations

- [x] TypeScript
- [x] Java
- [x] Ruby

### CODE REVIEW

**Comment by davidjgoss** [comment]
TypeScript and Java done - would anyone mind doing Ruby?

**Comment by aurelien-reeves** [comment]
> TypeScript and Java done - would anyone mind doing Ruby?

Will do

File: ruby/lib/cucumber/ci_environment.rb [hunk 1]
```diff
@@ -23,19 +23,28 @@ def detect(ci_environment, env)
...
     end
 
+    def detect_git(ci_environment, env)
+      revision = evaluate(ci_environment['git']['revision'], env)
+      return nil if revision.nil?
+
+      git_info = {
+        remote: remove_userinfo_from_url(evaluate(ci_environment['git']['remote'], env)),
+        revision: evaluate(ci_environment['git']['revision'], env),
+        branch: evaluate(ci_environment['git']['branch'], env),
```
Comment (aslakhellesoy) [inline ruby/lib/cucumber/ci_environment.rb:40]: I think we should conditionally assign `:branch` in the same way we conditionally assign `:tag`

File: ruby/lib/cucumber/ci_environment.rb [hunk 2]
```diff
@@ -23,19 +23,28 @@ def detect(ci_environment, env)
...
+
+      detected_git = detect_git(ci_environment, env)
+      result[:git] = detected_git if detected_git
       result
     end
 
+    def detect_git(ci_environment, env)
+      revision = evaluate(ci_environment['git']['revision'], env)
+      return nil if revision.nil?
+
```
Comment (aslakhellesoy) [inline ruby/lib/cucumber/ci_environment.rb:36]: I think we should return early here if remote is nil.

File: ruby/lib/cucumber/ci_environment.rb (hunk 1 above)
Comment (aslakhellesoy) [reply ruby/lib/cucumber/ci_environment.rb:40]: We might need 2 new tests:

- One where remote is set and revision is not
- One where revision is set and remote is not

**Review by aslakhellesoy — CHANGES_REQUESTED** [review]

File: ruby/lib/cucumber/ci_environment.rb (hunk 1 above)
Comment (aurelien-reeves) [reply ruby/lib/cucumber/ci_environment.rb:40]: > We might need 2 new tests:
> 
>     * One where remote is set and revision is not
> 
>     * One where revision is set and remote is not

What do we do in those cases? We do not set git even if one of the two data exists?

**Comment by aurelien-reeves** [comment]
I'll do the changes for ruby and js - or maybe David you would like to do it for js?
I'll take a look at the java impl too

File: ruby/lib/cucumber/ci_environment.rb (hunk 1 above)
Comment (aslakhellesoy) [reply ruby/lib/cucumber/ci_environment.rb:40]: Correct - see details in https://github.com/cucumber/ci-environment/issues/36

**Review by aslakhellesoy — APPROVED** [review]
Looks great!

**Comment by aslakhellesoy** [comment]
Great work! I've release 8.0.0

**Comment by aurelien-reeves** [comment]
I like the async team effort here 🤩

**Pull request merged**